### PR TITLE
tracing: instrument opentracing

### PR DESCRIPTION
**Pull request by abraithwaite**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -8,6 +8,7 @@ const debug = require('debug')('jsonrpc2')
 const split = require('split2')
 const once = require('once')
 const uid = require('uid2')
+const opentracing = require('opentracing')
```
Comment (nettofarah) [inline index.js]: I'm not sure if `opentracing` should be a dependency of this library.
I believe it should be passed as either:

a) a dependency (following a certain interface) that can be injected on instantiation time
b) a middleware that does the reporting

File: index.js [hunk 2]
```diff
@@ -31,14 +32,20 @@ class Client {
...
+
     this.logMiddleware = this.logMiddleware.bind(this)
     this.callMiddleware = this.callMiddleware.bind(this)
 
     this.rebuildMiddleware()
   }
 
   makeHTTPRequest (body, options, fn) {
-    const requestOptions = {
+    var requestOptions = {
```
Comment (nettofarah) [inline index.js]: we should default to `let` over `var`

File: index.js [hunk 3]
```diff
@@ -66,7 +74,14 @@ class Client {
 
       debug('success %s: %j', options.id, body.result || {})
       fn(null, body.result)
-    })
+    }
+
+    if (this.annotateTrace) {
+      var annotated = this.annotateTrace(body.method, requestOptions, callback)
```
Comment (nettofarah) [inline index.js]: it is a good practice to use `const` when you're not modifying the reference to a variable

File: index.js [hunk 4]
```diff
@@ -47,7 +54,8 @@ class Client {
       requestOptions.headers = { 'user-agent': this.userAgent }
     }
 
-    request.post(this.url, requestOptions, (err, res, body) => {
+    requestOptions.uri = this.url
+    var callback = function (err, res, body) {
```
Comment (nettofarah) [inline index.js]: I think `const` makes more sense here

File: index.js [hunk 5]
```diff
@@ -176,4 +191,33 @@ class Client {
   }
 }
 
+// tracing takes a config and returns a function with the signature
```
Comment (nettofarah) [inline index.js]: The comments are great here.
But it makes more sense to use JSDocs here, as it is common practice for oss js projects: http://usejsdoc.org/

**Comment by nettofarah** [comment]
Here's an example of how to use middlewares: https://github.com/segmentio/jsonrpc2.js#middleware

**Comment by nettofarah** [comment]
@abraithwaite feel free to ping me if you need any help with this. These are some exciting changes! 💃

File: index.js (hunk 1 above)
Comment (abraithwaite) [reply index.js]: I agree that it shouldn't be an explicit requirement here.  I was trying to think of other options, but I realized that it itself is an interface.  So if you call `globalTracer` it returns a no-op.

I looked briefly at writing a middleware but it didn't work out for some reason.  Lemme verify that again.

File: opentracing/index.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    if (options.span !== undefined) {
+      spanOpts = {childOf: options.span.context()}
+    }
+    const name = context.method || 'external_rpc'
+    const span = tracer.startSpan(name, spanOpts)
+    tracer.inject(span, opentracing.FORMAT_HTTP_HEADERS, options.headers)
+    options.span = span
+    context.options = options
+    return new Promise(resolve => {
+      span.finish()
```
Comment (abraithwaite) [inline opentracing/index.js]: uhh, do I put this before or after resolve?

File: index.js (hunk 2 above)
Comment (vadimdemedes) [reply index.js]: Correction: it should always be `const` first, then `let` and never `var`.

File: opentracing/index.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
+'use strict'
+
+const opentracing = require('opentracing')
+
+// tracing takes a config and returns a function with the signature
+// for koa middleware: function(context, next)
+function tracing (config) {
```
Comment (vadimdemedes) [inline opentracing/index.js]: Use arrow functions instead of "old" ones.

File: opentracing/index.js [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+  const conf = config || {}
+  const tracer = conf.tracer || opentracing.globalTracer()
+
+  return function (context, next) {
+    var options = context.options || {}
+    options.headers = options.headers || {}
+    var spanOpts = {}
+    if (options.span !== undefined) {
```
Comment (vadimdemedes) [inline opentracing/index.js]: => `if (options.span) {`

File: opentracing/index.js [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+      spanOpts = {childOf: options.span.context()}
+    }
+    const name = context.method || 'external_rpc'
+    const span = tracer.startSpan(name, spanOpts)
+    tracer.inject(span, opentracing.FORMAT_HTTP_HEADERS, options.headers)
```
Comment (vadimdemedes) [inline opentracing/index.js]: Could you add some spacing here, between variable declarations and fn calls?

File: opentracing/index.js [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    options.span = span
+    context.options = options
+    return new Promise(resolve => {
+      span.finish()
+      resolve(next())
```
Comment (vadimdemedes) [inline opentracing/index.js]: You could just call `next()` and return it, without `new Promise`.

File: test/http.js [hunk 11]
```diff
@@ -62,6 +63,26 @@ test('send userAgent in request header', async t => {
   t.is(res['user-agent'], 'test/1.0')
 })
 
+test('tracer wrapper', async t => {
+  var started = false
```
Comment (vadimdemedes) [inline test/http.js]: Use sinon instead of boolean variables to check if a fn was called.

File: test/http.js [hunk 12]
```diff
@@ -62,6 +63,26 @@ test('send userAgent in request header', async t => {
...
+  var injected = false
+  var finished = false
+  var client = new Client(address)
```
Comment (vadimdemedes) [inline test/http.js]: => `const`

File: test/http.js [hunk 13]
```diff
@@ -62,6 +63,26 @@ test('send userAgent in request header', async t => {
...
+  client.use(tracing({ tracer: {
```
Comment (vadimdemedes) [inline test/http.js]: =>
```js
client.use(tracing({
  tracer: {
    ...
  }
})
```

**Comment by abraithwaite** [comment]
Sorry I should have closed this with the other PR.

Thanks for the feedback though!

**Comment by vadimdemedes** [comment]
Hey @abraithwaite, could you fix the issues above in a separate PR?

**Comment by abraithwaite** [comment]
Which issues?  I've closed this PR as we decided to go with using middleware.

**Comment by vadimdemedes** [comment]
Ah, my bad, I thought it was merged ;)

**Pull request closed without merging**